### PR TITLE
Update `swanky check`

### PR DESCRIPTION
**Pull request by prxgr4mm3r**

### CODE REVIEW

File: src/commands/check/index.ts [hunk 1]
```diff
@@ -98,7 +121,19 @@ export default class Check extends SwankyCommand<typeof Check> {
...
-          const supportedInk = ctx.swankyConfig?.node.supportedInk;
+          let supportedInk = ctx.swankyConfig?.node.supportedInk;
+          const regex = /cargo-contract-contract (.*)-unknown-x86_64-unknown-linux-gnu/;
+          const cargoContract = ctx.versions.tools.cargoContract;
+          if (cargoContract) {
+            const match = cargoContract.match(regex);
+            if (match) {
+              const cargoVersion = match[1];
+              const version = cargoContractDeps.get(cargoVersion);
+              if (version && semver.gt(supportedInk!, version[0])) {
```
Comment (ipapandinas) [inline src/commands/check/index.ts:132]: Checking order is wrong here, version goes first. Also the higher version should be the one supported:

`semver.gt(versions[versions.length - 1], supportedInk!)`

File: src/commands/check/index.ts [hunk 2]
```diff
@@ -98,7 +121,19 @@ export default class Check extends SwankyCommand<typeof Check> {
...
         task: async (ctx) => {
-          const supportedInk = ctx.swankyConfig?.node.supportedInk;
+          let supportedInk = ctx.swankyConfig?.node.supportedInk;
+          const regex = /cargo-contract-contract (.*)-unknown-x86_64-unknown-linux-gnu/;
+          const cargoContract = ctx.versions.tools.cargoContract;
+          if (cargoContract) {
+            const match = cargoContract.match(regex);
+            if (match) {
+              const cargoVersion = match[1];
+              const version = cargoContractDeps.get(cargoVersion);
```
Comment (ipapandinas) [inline src/commands/check/index.ts:131]: Make it plural: `versions`

File: src/commands/check/index.ts [hunk 3]
```diff
@@ -118,12 +153,27 @@ export default class Check extends SwankyCommand<typeof Check> {
           ctx.mismatchedVersions = mismatched;
         },
       },
+      {
+        title: "Check for missing tools",
```
Comment (ipapandinas) [inline src/commands/check/index.ts:157]: I uninstalled `cargo-dylint` to test this one. Here is my output:

```bash
➜  temp_project git:(main) ✗ ../bin/run.js check
✔ Check OS
✔ Check Rust
✔ Check cargo
✔ Check cargo nightly
✔ Check cargo dylint
✔ Check cargo-contract
✔ Check swanky node
✔ Read ink dependencies
✔ Verify ink version
✔ Check for missing tools

{ platform: 'darwin', architecture: 'arm64' } {
  tools: {
    rust: 'rustc 1.72.1 (d5c2e9c34 2023-09-13)',
    cargo: 'cargo 1.72.1 (103a7ff2e 2023-08-15)',
    cargoNightly: 'cargo 1.77.0-nightly (2ce45605d 2024-01-04)',
    cargoDylint: null,
    cargoContract: 'cargo-contract-contract 4.0.0-rc.1-unknown-aarch64-apple-darwin'
  },
  missingTools: [ 'cargoDylint' ],
  contracts: { flipper: { ink: '4.2.1' } },
  node: '1.6.0'
}
```

`cargoDylint` is correctly added to the missing tools, however `✔ Check cargo dylint` as a success tick. Use a "X" tick instead

File: src/commands/check/index.ts [hunk 4]
```diff
@@ -133,9 +183,26 @@ export default class Check extends SwankyCommand<typeof Check> {
...
+    const supportedArch = ["arm64", "x64"];
+
+    if (!supportedPlatforms.includes(context.os.platform)) {
+      console.error(`[ERROR] Platform ${context.os.platform} is not supported`);
+    }
+    if (!supportedArch.includes(context.os.architecture)) {
+      console.error(`[ERROR] Architecture ${context.os.architecture} is not supported`);
+    }
+
+    console.log(context.os, context.versions);
```
Comment (ipapandinas) [inline src/commands/check/index.ts:196]: Make it 2 distincts console.log for newline purposes.

File: src/commands/check/index.ts [hunk 5]
```diff
@@ -133,9 +183,26 @@ export default class Check extends SwankyCommand<typeof Check> {
       Please use "=" to install a fixed version (Example: "=3.0.1")
       `);
     }
+    const supportedPlatforms = ["darwin", "linux"];
```
Comment (ipapandinas) [inline src/commands/check/index.ts:186]: OS checks should be moved to the "Check OS" first task

File: src/commands/check/index.ts [hunk 6]
```diff
@@ -118,12 +153,27 @@ export default class Check extends SwankyCommand<typeof Check> {
...
+          for (const [toolName, toolVersion] of Object.entries(ctx.versions.tools)) {
+            if (!toolVersion) {
+              missingTools.push(toolName);
+            }
+          }
+          ctx.versions.missingTools = missingTools;
+        },
+      }
     ]);
+    process.platform
```
Comment (ipapandinas) [inline src/commands/check/index.ts:169]: I think those processes are useless.

File: src/commands/check/index.ts [hunk 7]
```diff
@@ -98,7 +121,19 @@ export default class Check extends SwankyCommand<typeof Check> {
...
+            if (match) {
+              const cargoVersion = match[1];
+              const version = cargoContractDeps.get(cargoVersion);
+              if (version && semver.gt(supportedInk!, version[0])) {
+                supportedInk = version[0];
+              }
+            }
+          }
 
           const mismatched: Record<string, string> = {};
```
Comment (ipapandinas) [inline src/commands/check/index.ts:138]: Make it another task with title: "Verify ink dependencies"

**Review by ipapandinas — COMMENTED** [review]
Checkbox 4 from the original issue #114 is not fully completed:

>  fetch and store the dependency table on init and check use

We should update `node.supportedInk` version in swanky config when performing the `checkCliDependencies` task in the `init` command. You can make and helper performing the `semver.gt` against our cargo-contract dependencies map like you did in "Verify ink version" from the `check` command. Then update the `node.supportedInk` version in swanky config accordingly.

**Comment by prxgr4mm3r** [comment]
> Checkbox 4 from the original issue #114 is not fully completed:
> 
> > fetch and store the dependency table on init and check use
> 
> We should update `node.supportedInk` version in swanky config when performing the `checkCliDependencies` task in the `init` command. You can make and helper performing the `semver.gt` against our cargo-contract dependencies map like you did in "Verify ink version" from the `check` command. Then update the `node.supportedInk` version in swanky config accordingly.

Sorry, I don't understand properly what you mean by updating `node.supportedInk`. I thought that it should be set initially in `nodeInfo` for every SwankyNode version. Can you explain what factors affect the update of the `node.supportedInk`?

File: src/commands/check/index.ts [hunk 8]
```diff
@@ -17,6 +24,8 @@ interface Ctx {
       cargoDylint?: string | null;
       cargoContract?: string | null;
     };
+    supportedInk?: string;
+    missingTools: string[];
     contracts: Record<string, Record<string, string>>;
     node?: string | null;
```
Comment (ipapandinas) [inline src/commands/check/index.ts:30]: Name it swankyNode

**Review by ipapandinas — CHANGES_REQUESTED** [review]
@prxgr4mm3r  I reviewed the UX of the `check` command, the following tasks have to be perform in this order:

1. "Check OS"
2. "Check Rust"
3. "Check cargo"
4. "Check cargo nightly"
5. "Check cargo dylint"
6. "Check cargo-contract"
7. "Read ink dependencies"
8. "Verify ink version"
    (skipped if no swanky node is installed in swankyConfig)
    ensure compatibility with swanky node: the supportedInk should be greater or equal to the read ink version in step 7
    if not ensured => throw a WARN: `Version of ${inkPackage} (${version}) in ${contract} is higher than supported ink version (${supportedInk})`
9. "Verify cargo contract version"
  (skipped if no cargo contract installed => tools.cargoContract === null)
  ensure cargo contract version is greater or equal to the minimal version for the read ink version in step 7
10. "Check for missing tools"
  print a console.error(`[ERROR] ${error}`) for each missing tool

So "Check swanky node" becomes useless.

Also additional enhancements:

#### Try to print versions next to task success messages like this:

✔ Check OS: 'darwin-arm64'
✔ Check Rust: '1.72.1'
✔ Check cargo: '1.72.1'
...
You can extracted them using regex like you did for cargo contract

#### Also try to throw a WARN instead of error for dylint

✖ Cargo dylint is not installed!

#### Also checkbox 2 is not implemented

> add optional file output

It can be an optional flag `--print` and the output file should look like this:

```typescript
{
  "platform": "darwin",
  "architecture": "arm64",
  "tools": {
    "rust": "1.72.1",
    "cargo": "1.72.1",
    "cargoNightly": "1.77.0",
    "cargoContract": "4.0.0-rc.1"
  },
  "missingTools": [],
  "contracts": { "flipper": { "ink": "4.2.1" } },
  "swankyNode": null
}
```

File: src/lib/cargoContractInfo.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+export type cargoContractInfo = typeof cargoContractDeps;
+
+export const cargoContractDeps  = new Map([
```
Comment (ipapandinas) [inline src/lib/cargoContractInfo.ts:3]: The deps object should be adjusted like this:

```
const CARGO_CONTRACT_INK_DEPS: CargoContractInkDependency[] = [
  { minCargoContractVersion: "2.0.0", inkVersion: "4.0.0" },
  { minCargoContractVersion: "2.0.2", inkVersion: "4.0.1" },
  { minCargoContractVersion: "2.2.0", inkVersion: "4.2.0" },
  { minCargoContractVersion: "4.0.0", inkVersion: "5.0.0" },
];
```

File: src/commands/check/index.ts [hunk 10]
```diff
@@ -97,16 +162,34 @@ export default class Check extends SwankyCommand<typeof Check> {
       },
       {
         title: "Verify ink version",
```
Comment (ipapandinas) [inline src/commands/check/index.ts:164]: This becomes "Verify cargo contract version". The goal is to ensure a min cargo contract version is installed for the higher ink version used in contracts. This is ensured with `CARGO_CONTRACT_INK_DEPS`.

I think you can do something like this with semver:

```
const compatibleInkVersion = cargoContractInkDependencies.find(dep =>
        semver.satisfies(cargoContractVersion, `>=${dep.minCargoContractVersion}`)
      )?.inkVersion;
```

**Review by ipapandinas — CHANGES_REQUESTED** [review]

**Review by ipapandinas — APPROVED** [review]

File: src/lib/cargoContractInfo.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+export interface CargoContractInkDependency {
+  minCargoContractVersion: string;
+  validInkVersionRange: string;
+}
+
+// Keep cargo-contract versions in descending order
+// Ranges are supported by semver
+export const CARGO_CONTRACT_INK_DEPS: CargoContractInkDependency[] = [
+  { minCargoContractVersion: "4.0.0", validInkVersionRange: "<99.0.0" }, // Non-max version known yet: a very high version is used as fallback in the meantime
+  { minCargoContractVersion: "2.2.0", validInkVersionRange: "<5.0.0" },
```
Comment (pmikolajczyk41) [inline src/lib/cargoContractInfo.ts:10]: why do you skip `3.x` versions?

File: src/commands/check/index.ts [hunk 12]
```diff
@@ -17,57 +26,117 @@ interface Ctx {
...
+  mismatchedVersions: Record<string, string>;
   looseDefinitionDetected: boolean;
 }
 
 export default class Check extends SwankyCommand<typeof Check> {
   static description = "Check installed package versions and compatibility";
 
+  static flags = {
+    print: Flags.string({
+      char: "p",
```
Comment (pmikolajczyk41) [inline src/commands/check/index.ts:44]: why `p`? 'path'? I guess 'o' is more common

File: src/commands/check/index.ts [hunk 13]
```diff
@@ -17,57 +26,117 @@ interface Ctx {
...
+          ctx.versions.tools.cargoNightly = (await commandStdoutOrNull("cargo +nightly -V"))?.match(/cargo (.*)-nightly \((.*)/)?.[1];
+          if (!ctx.versions.tools.cargoNightly) {
+            throw new Error("Cargo nightly is not installed");
+          }
+          task.title = `Check cargo nightly: ${ctx.versions.tools.cargoNightly}`;
         },
+        exitOnError: false,
       },
       {
         title: "Check cargo dylint",
```
Comment (pmikolajczyk41) [inline src/commands/check/index.ts:108]: where `dylint` is actually used when working with swanky?

**Review by pmikolajczyk41 — APPROVED** [review]

File: src/lib/cargoContractInfo.ts (hunk 11 above)
Comment (ipapandinas) [reply src/lib/cargoContractInfo.ts:10]: According to the latest change logs, [cargo-contract `v2.2.0`](https://github.com/paritytech/cargo-contract/releases/tag/v2.2.0) specifies updating ink! to a minimum version of `v4.2.0`. While cargo-contract versions `v3.x` remain supported, they do not introduce new minimum requirements for ink!, thus maintaining the requirement for ink! `v4.2.0`.

File: src/commands/check/index.ts (hunk 13 above)
Comment (ipapandinas) [reply src/commands/check/index.ts:108]: Nowhere directly, `dylint` has been in place for some time for smart contract projects using swanky. I believe it aligns with the recommended best practices for installation as outlined in the [cargo-contract README](https://github.com/paritytech/cargo-contract/blob/744babfcd61fd74d2e02d20b25b9823ecf18308a/README.md?plain=1#L45).

File: src/commands/check/index.ts (hunk 12 above)
Comment (ipapandinas) [reply src/commands/check/index.ts:44]: you're right, bienvu!

**Comment by pmikolajczyk41** [comment]
why was it closed instead of merged? @ipapandinas

**Comment by ipapandinas** [comment]
I created a single PR #205 to address the overlap between features and to prepare for the upcoming release. I closed this particular one due to significant conflicts. @pmikolajczyk41

**Comment by pmikolajczyk41** [comment]
@ipapandinas okay, I was just confused that the other 3 PRs were merged, so I wanted to ensure - thanks!

**Pull request closed without merging**